docker_push_latest_if_changed.py

```diff
@@ -1,5 +1,120 @@
 #!/usr/bin/env python3
 import argparse
+import hashlib
+from collections import namedtuple
+
+import docker_util
+from verbose_print import print_v
+from verbose_print import print_vv
+from verbose_print import set_print_level
+
+
+ImageParityFields = namedtuple(
+    'ImageParityFields',
+    ('image_cmds_hash', 'packages_hash'),
+)
+
+GET_PACKAGE_LIST_COMMAND = ['dpkg', '-l']
+
+
+def _docker_push_latest_if_changed(source, target, is_dry_run):
+    _validate_source(source)
+    target = _get_sanitized_target(source, target)
+
+    print_v('Pushing source image')
+    docker_util.push_image(source, is_dry_run)
+    try:
+        print('Pulling target image...')
+        docker_util.pull_image(target)
+    except docker_util.ImageNotFoundError:
+        print((
+            f'Target image {target} was not found in the registry. '
+            'Going to attempt to tag and push the target image anyway.'
+        ))
+        docker_util.tag_image(source, target, is_dry_run)
+        docker_util.push_image(target, is_dry_run)
+    else:
+        if _is_image_changed(source, target):
+            print((
+                'Source image was found to be different from the current '
+                'version of the target image. Tagging and pushing new target '
+                'image.'
+            ))
+            docker_util.tag_image(source, target, is_dry_run)
+            docker_util.push_image(target, is_dry_run)
+        else:
+            print((
+                'Source image was found to be the same as the current '
+                'version of the target image. Not pushing the target image.'
+            ))
+
+
+def _validate_source(source):
+    print_vv('Docker inspecting source to verify it exists')
+    source_inspect = docker_util.inspect_image(source)
+    if source not in source_inspect['RepoTags']:
```

Comment:
copying this from above, mostly because github ui can suck:

```
ah! in that case just use `urllib.parse.urlparse` -- they _are urls_ after all :)

this'll remove the ambiguity about port I believe and give you some confidence on whether they're tagged or not
```

Reply:
If the source was `127.0.0.1:12345/coolimage:lol` the urlparse result would be:
`ParseResult(scheme='127.0.0.1', path='1235/coolimage:lol')`
To which we can check if there's a colon in path and things would be fine

However if say the source image omitted the registry portion and it was `coolimage:oldest` urlparse would give:
`ParseResult(scheme='coolimage', path='oldest', ...)`
A colon check would see this as an invalid source, but this is actually a repo of name `coolimage` and tag `oldest`

We could probably assume with pretty good confidence that the registry part will be included, and it'll be simple to validate that it's there with urlparse. If someone is looking for some reason to push to the public registry though, they'll need to remember to prefix with `registry.hub.docker.com/library`. We'll probably never need such functionality though